test/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/supportingevidence/SupportingEvidenceControllerSpec.scala

```diff
@@ -383,7 +383,299 @@ class SupportingEvidenceControllerSpec
       }
     }
 
-    //TODO: test submission
+    "handling requests to chose evidence document type" must {
+
+      def performAction(uploadReference: UploadReference)(data: Seq[(String, String)]): Future[Result] =
+        controller.chooseSupportingEvidenceDocumentTypeSubmit(uploadReference)(FakeRequest().withFormUrlEncodedBody(data: _*))
+
+      "fail" when {
+        "document type is missing" in {
+          val uploadReference = sample[UploadReference]
+          val answers = IncompleteSupportingEvidenceAnswer(evidences = List.empty)
+
+          val (session, _, _) = sessionWithClaimState(Some(answers))
+
+          inSequence {
+            mockAuthWithNoRetrievals()
+            mockGetSession(session)
+          }
+
+          status(performAction(uploadReference)(Seq.empty)) shouldBe BAD_REQUEST
+        }
+
+        "supporting evidence is missing for incomplete journey" in {
+          the[RuntimeException] thrownBy {
+            val uploadReference = sample[UploadReference]
+            val answers = IncompleteSupportingEvidenceAnswer(evidences = List.empty)
+
+            val (session, _, _) = sessionWithClaimState(Some(answers))
+
+            inSequence {
+              mockAuthWithNoRetrievals()
+              mockGetSession(session)
+            }
+
+            await(performAction(uploadReference)(Seq(SupportingEvidenceController.chooseDocumentTypeDataKey -> "0")))
+          } should have message "could not find uploaded file"
+        }
+
+        "supporting evidence is missing for complete journey" in {
+          val uploadReference = sample[UploadReference]
+          val answers = CompleteSupportingEvidenceAnswer(evidences = List.empty)
+
+          val (session, _, _) = sessionWithClaimState(Some(answers))
+
+          inSequence {
+            mockAuthWithNoRetrievals()
+            mockGetSession(session)
+          }
+
+          val caught = intercept[RuntimeException] {
+            await(performAction(uploadReference)(Seq(SupportingEvidenceController.chooseDocumentTypeDataKey -> "0")))
+          }
+
+          caught.getMessage shouldBe s"could not find file upload with reference: $uploadReference"
+        }
+
+        "caught an error on session update" in {
+          val uploadReference = sample[UploadReference]
+          val uploadRequest = sample[UploadRequest]
+          val upscanUploadMeta = UpscanUploadMeta(
+            uploadReference.value,
+            uploadRequest
+          )
+          val upscanSuccess = sample[UpscanSuccess]
+
+          val supportingEvidence = SupportingEvidence(
```

Comment:
it would nice to have a generator for this class - if I remember correctly I had a problem creating a generator for it as the type had a sealed trait hierarchy as the type for one of it's fields but I think I fixed that problem now.